File: problems/discnt/solution/java/src/Discnt.java
import java.io.*;
import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;
import java.util.Scanner;

public class Discnt {

    public static void main(String[] args) throws IOException {
        String inputFileName = args.length >= 2 ? args[0] : "discnt.in";
        String outputFileName = args.length >= 2 ? args[1] : "discnt.out";

        DiscntInputData inputData = readInput(inputFileName);
        DiscntOutputData outputData = solve(inputData);
        writeOutput(outputFileName, outputData);
    }

    private static DiscntInputData readInput(String inputFileName) throws FileNotFoundException {
        File inputFile = new File(inputFileName);

        try (Scanner inputFileScanner = new Scanner(inputFile)) {
            String[] priceStrings = inputFileScanner.nextLine().split(" ");

            int[] prices = new int[priceStrings.length];
            for (int i = 0; i < priceStrings.length; i++) {
                prices[i] = Integer.parseInt(priceStrings[i]);
            }

            int discountPercentage = inputFileScanner.nextInt();

            return new DiscntInputData(prices, discountPercentage);
        }
    }

    private static DiscntOutputData solve(DiscntInputData inputData) {
        int[] prices = inputData.getPrices();
        int numberOfDiscountedItems = prices.length / 3;

        // Sort by price in descending order.
        MergeSort.sort(prices);

        double minPurchaseSum = 0;
        for (int i = 0; i < prices.length; i++) {
            if (i < numberOfDiscountedItems) {
                minPurchaseSum += prices[i] * ((100.0 - inputData.getDiscountPercentage()) / 100);
            }
            else {
                minPurchaseSum += prices[i];
            }
        }

        return new DiscntOutputData(minPurchaseSum);
    }

    private static void writeOutput(String outputFileName, DiscntOutputData outputData) throws IOException {
        DecimalFormatSymbols numberFormatSymbols = new DecimalFormatSymbols();
        numberFormatSymbols.setDecimalSeparator('.');
        DecimalFormat numberFormat = new DecimalFormat("0.00", numberFormatSymbols);
        numberFormat.setGroupingUsed(false);

        try (Writer outputFileWriter = new FileWriter(outputFileName)) {
            outputFileWriter.write(numberFormat.format(outputData.getMinPurchaseSum()));
        }
    }
}

File: problems/discnt/solution/python/discnt.py
import sys


def main():
    input_filename = "discnt.in" if len(sys.argv) == 1 else sys.argv[1]
    output_filename = "discnt.out" if len(sys.argv) == 1 else sys.argv[2]

    prices, discount_rate = read_input(input_filename)
    min_purchase_sum = solve(prices, discount_rate)
    write_output(output_filename, min_purchase_sum)


def read_input(filename):
    with open(filename, "r") as input_file:
        prices = [int(price_str) for price_str in input_file.readline().split()]
        discount_rate = int(input_file.readline())
        return prices, discount_rate


def solve(prices, discount_rate):
    insertion_sort(prices)
    num_discounted_items = len(prices) / 3
    for i in range(0, num_discounted_items):
        prices[i] *= ((100.0 - discount_rate) / 100)
    return sum(prices)


def insertion_sort(array):
    for i in xrange(1, len(array)):
        j = i
        # Sort in descending order
        while j > 0 and array[j - 1] < array[j]:
            array[j - 1], array[j] = array[j], array[j - 1]
            j -= 1


def write_output(filename, min_purchase_sum):
    with open(filename, "w") as output_file:
        output_file.write("%.2f" % min_purchase_sum)


if __name__ == "__main__":
    main()
